parquet/src/main/java/org/apache/iceberg/parquet/TypeWithSchemaVisitor.java

```diff
@@ -64,7 +64,7 @@ public static <T> T visit(
       } else if (annotation instanceof LogicalTypeAnnotation.VariantLogicalTypeAnnotation
           || (iType != null && iType.isVariantType())) {
         // when Parquet has a VARIANT logical type, use it here
-        return visitVariant(iType.asVariantType(), group, visitor);
+        return visitVariant(iType != null ? iType.asVariantType() : null, group, visitor);
```

Comment:
So this NPE issue would happen when we prune out a variant column that we know we don't need to read; and in that case the iceberg type passed to the visitor would expectedly be null. Makes sense, it's really no different than the other cases when pruning.